### DISCUSSION TITLE
KEMC model

### DISCUSSION THREAD

**Discussion by MaloValmalle** [thelfer/tfel#316] [Q&A] [answered]
Hi everyone,

I was wondering if someone has already tried to implement the KEMC model using Mfront?

Since I'm new to implementing constitutive laws I'll prefer not to start from scratch.
The main difficulties I'm having with this particular model are :

-the introduction of a new state variable
-the definition of the equivalent plastic strain rate

If someone has some tips that would be wonderful!

I've put an article using this constitutive law in the attachments if you want to have a look (section 3).

Cheers to everyone,

Malo
[386-belotteau09.pdf](https://github.com/thelfer/tfel/files/9584839/386-belotteau09.pdf)

**Comment by thelfer** [comment, marked as answer]
@MaloValmalle Please find attach a first draft
[KEMC.zip](https://github.com/thelfer/tfel/files/9585248/KEMC.zip). Absolutely no guarantee, but probably a good starting point.
It needs to be modified with the real material parameter, then tested with `MTest` before any FEM computation and finally we could write the analytical jacobian.

**Reply by MaloValmalle** [reply]
Thanks a lot @thelfer , I'll try and keep you updated!

**Reply by MaloValmalle** [reply]
Hi @thelfer 

As it was not working at first I've tried to modify it this weekend! All the equations seemed to be ok except for a power that was missing. 

I've modified the first draft in order to have an elastic loading and then a plastic loading, using the Johnson Cook (attached) as a model. 

When I run the test it is always fine till the plastic loading occurs, I'm wondering if it is because the yield surface definition is implicit?
[KEMC.zip](https://github.com/thelfer/tfel/files/9596813/KEMC.zip)

Also because the yield surface definition is implicit I'm not sure if the criterion that I've set to enter plasticity is ok (simple yield stress)....

I've got no clue where the problem could be, if you could have a look that would be very nice of you !

**Reply by thelfer** [reply]
@MaloValmalle Your changes seem strange and I don't have time to review them (in particular the dp_mts variable seems dubious). I however did a mistake since I should have defined `f` as `max(seq-R, stress(0))`. I corrected this mistake and imported your material parameters in my own implementation (see the attached file).

However, there are probably inconsistencies since the `s0` is of the order of 1e-18, which makes the `sinh` explode as expected. You probably shall check every material parameters in the first place as follows:

```c++
@InitLocalVariables {
  constexpr auto kb = PhysicalConstants::kb;
  const auto T_mts = T + theta * dT;
  de = de0 * exp(-Ea / (kb * (T_mts)));
  s0 = kb * T_mts / Va;
  std::cout << "de: " << de << std::endl;
  std::cout << "s0: " << s0 << std::endl;
  std::exit(-1);
}
```
[KEMC.zip](https://github.com/thelfer/tfel/files/9597092/KEMC.zip)

**Reply by MaloValmalle** [reply]
Hi @thelfer,

Thanks a lot for your help!

I had indeed some problems with the constants.
I also had some underflow errors with the de (because of the really small constants it was equal to 0). 
I fixed it by writing the sinh with exponentials and calculating everything at the end.

Everything seems to work now. Here is the code + a fig showing a behavior that is qualitatively in agreement with the results of the article.

![model_300K](https://user-images.githubusercontent.com/90772227/191053009-51d1fa69-0f84-488d-9a6d-6f803152f4d3.jpg)
[KEMC.zip](https://github.com/thelfer/tfel/files/9600298/KEMC.zip)

Thanks a lot for your precious help!

Br,

Malo

**Reply by MaloValmalle** [reply]
(strain rates in legend/ simulation at 300K)

**Reply by thelfer** [reply]
@MaloValmalle There seems something strange in our code: the activation energy is not divided by k*T in the energy. To be honest, I would not merge the two exponentials. Are you sure that Ea and f/Va are of the same magnitude ? One must always try to avoid cancellation errors...

**Reply by MaloValmalle** [reply]
Hi, 
Once again I did make a stupid mistake, I corrected that and found that the error was indeed Ea not being converted to SI units...
With the good set of parameters everything seems to work well!
[KEMC.zip](https://github.com/thelfer/tfel/files/9605567/KEMC.zip)
(file containing the previous code and the modified code with the good set of parameters)
Thanks for your guidance I did not think it could be as tricky beforehand ;)
